## Assembly

### Directory structure
Nanopore raw data directory or basecalls and paired-end short reads are expected as input. Data are expected to be basecalled with Guppy 6.2.1 SUP although this is modifiable in the Snakefile.  
  
Nanopore: raw data in `./{species}` *or* basecalls as `{species}.passReads.guppy621.fastq.gz`  
Illumina: `./{species}_1.fastq.gz` `./{species}_2.fastq.gz`  
  
output soft-masked genome: `{species}.rm.fasta`  
repeat library: `{species}-families.fa`  
repeat tracks: `{species}.out.gff`  

### Sherlock cmd
```bash
snakemake --cores 16 --jobs 30 --use-singularity --rerun-incomplete \
    --cluster "sbatch -p {cluster.partition} --mem {cluster.mem} -c {cluster.cpus-per-task} -t {cluster.time} --constraint {cluster.constraint} --gpus {cluster.gpus}" \
    --cluster-config cluster_config.yml --singularity-args ' --nv'
```

## Variant calling

### Directory structure
Data files are expected to be in the following directory/file structure. The rest are generated by Snakemake.

reference genome: `data/{species}.fa`  
short reads: `data/samples/{species}/*_1.fastq.gz` `data/samples/{species}/*_2.fastq.gz`  
  
output VCFs: `vcfs/{species}/`  
output mappability track: `{species}.GenMap.bedgraph`

### Sherlock cmd
```bash
species="D.melanogaster" snakemake --cores 16 --jobs 30 --use-singularity --rerun-incomplete \
    --cluster "sbatch -p {cluster.partition} --mem {cluster.mem} -c {cluster.cpus-per-task} -t 24:00:00" \
    --cluster-config cluster_config.yml
```
