## Genome assembly

The genome assembly pipeline is run with a single command:

```bash
snakemake --cores $(nproc) --jobs $(nproc) --use-singularity \
  --rerun--incomplete --config sample="D.melanogaster" \
  --config fwd="D.melanogaster_R1.fastq.gz" \
  --config rev="D.melanogaster_R2.fastq.gz"
```

The Snakemake workflow makes some assumptions about the directory structure and 

###


### Input files
For each genome assembly sample (`${sample}`), the following data are expected in the same working 
directory as the `Snakefile`:
* Nanopore raw data (`./${sample}/` folder containing fast5/pod5 files) *or*  gzipped basecalls (`${sample}.fastq.gz`)
* Illumina paired-end reads (`${sample}_R1.fastq.gz`,`${sample}_R2.fastq.gz`)

### Output files  
* output soft-masked genome: `{species}.rm.fasta`  
* repeat library: `{species}-families.fa`  
* repeat tracks: `{species}.out.gff`  

### Sherlock cmd
```bash
snakemake --cores 16 --jobs 30 --use-singularity --rerun-incomplete \
    --cluster "sbatch -p {cluster.partition} --mem {cluster.mem} -c {cluster.cpus-per-task} -t {cluster.time} --constraint {cluster.constraint} --gpus {cluster.gpus}" \
    --cluster-config cluster_config.yml --singularity-args ' --nv'
```

## Variant calling

### Directory structure
Data files are expected to be in the following directory/file structure. The rest are generated by Snakemake.

reference genome: `data/{species}.fa`  
short reads: `data/samples/{species}/*_1.fastq.gz` `data/samples/{species}/*_2.fastq.gz`  
  
output VCFs: `vcfs/{species}/`  
output mappability track: `{species}.GenMap.bedgraph`

### Sherlock cmd
```bash
species="D.melanogaster" snakemake --cores 16 --jobs 30 --use-singularity --rerun-incomplete \
    --cluster "sbatch -p {cluster.partition} --mem {cluster.mem} -c {cluster.cpus-per-task} -t 24:00:00" \
    --cluster-config cluster_config.yml
```
